### PR TITLE
Application

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @gkoz @sdroege

### CODE REVIEW

File: src/bin/basic.rs [hunk 1]
```diff
@@ -1,35 +1,64 @@
...
-}
\ No newline at end of file
+}
+
+fn main() {
+    let application = gtk::Application::new("com.github.basic",
+                                            gio::ApplicationFlags::empty())
+                                       .expect("Initialization failed...");
+
+    application.connect_startup(move |app| {
```
Comment (EPashkin) [inline src/bin/basic.rs]: `move` unneeded here

**Comment by EPashkin** [comment]
Only  problem with be that on windows, that it created listeners for `tcp://[::1] Port 54038` (port changes)
by running `"C:\WINDOWS\system32\rundll32.exe" D:\P\msys64\mingw64\bin\LIBGIO~1.DLL,g_win32_run_session_bus`.
If firewall disallow create listeners, then program stall for ~40sec.

File: src/bin/basic.rs [hunk 2]
```diff
@@ -1,35 +1,64 @@
...
 
-fn main() {
-    if gtk::init().is_err() {
-        println!("Failed to initialize GTK.");
-        return;
-    }
+use std::env::args;
+
+// make moving clones into closures more convenient
+macro_rules! clone {
```
Comment (EPashkin) [inline src/bin/basic.rs]: Seems we need add this macro to glib, or use separate crate (I seen it in this month but can't remember when).

**Comment by EPashkin** [comment]
Also maybe this part better added to `gtk::Window` as `connect_application_and_show_all` ?
Of course without usage of `clone!`.
```
        window.set_application(application);
        window.connect_delete_event(clone!(window => move |_, _| {
            window.destroy();
            Inhibit(false)
        }));
        window.show_all();
```
PS. `window.set_application(application);` not present in `basic.rs` or some others.

**Comment by sdroege** [comment]
> Only problem with be that on windows, that it created listeners for tcp://[::1] Port 54038 (port changes)
by running "C:\WINDOWS\system32\rundll32.exe" D:\P\msys64\mingw64\bin\LIBGIO~1.DLL,g_win32_run_session_bus.
If firewall disallow create listeners, then program stall for ~40sec.

Please file a bug against GTK for that, that seems unacceptable

File: src/bin/basic.rs (hunk 2 above)
Comment (sdroege) [reply src/bin/basic.rs]: https://crates.io/crates/closet

**Comment by sdroege** [comment]
> Also maybe this part better added to gtk::Window as connect_application_and_show_all ?

While it's used like this often in the example, it could very well be different in real applications.

File: src/bin/basic.rs (hunk 2 above)
Comment (EPashkin) [reply src/bin/basic.rs]: Yes, thanks.

**Comment by EPashkin** [comment]
@sdroege you right about `connect_application_and_show_all`. Normal application may need do some work on exiting.


About listener:
it caused by nonexistence build-in DBus for Windows.
If any example running, others don't create new listeners and use first's.
So IMHO nothing to report.

File: src/bin/basic.rs (hunk 2 above)
Comment (GuillaumeGomez) [reply src/bin/basic.rs]: I was just thinking adding a new crate to `gtk-rs` with some utils like that.

**Comment by GuillaumeGomez** [comment]
> PS. `window.set_application(application);` not present in `basic.rs` or some others.

`set_application` is useful only for glade examples in order to set the setup `application` to the `ApplicationWindow`. Otherwise it's useless.

**Comment by GuillaumeGomez** [comment]
I'm merging it then.

**Pull request merged**